### ISSUE TITLE
Maybe some bug in multi gpu training

### ISSUE THREAD

**Issue by Kfkcome** [huggingface/trl#2960]
### Reproduction

I encountered some issues while training my Qwen2.5 1.5B model using the GPROTrainer from the trl library. My hardware environment is A800 x4 80G GPUs.

**First Attempt (Single A800):**  
Command:
```bash
accelerate launch --config_file recipes/accelerate_configs/zero3.yaml --num_processes=1 src/open_r1/grpo.py \
    --output_dir DeepSeek-R1-Distill-Qwen-15B-GRPO-new \
    --model_name_or_path /data1/gaoxingjie/model/Qwen2.5-1.5B-Instruct/ \
    --dataset_name data/gsm8k/main \
    --max_prompt_length 512 \
    --max_completion_length 1024 \
    --per_device_train_batch_size 8 \
    --num_generations 8 \
    --lr_scheduler_type cosine \
    --warmup_ratio 0.1 \
    --logging_strategy steps \
    --learning_rate 3.0e-06 \
    --gradient_accumulation_steps 16 \
    --logging_steps 5 \
    --eval_strategy no \
    --bf16 \
    --use_vllm \
    --vllm_device auto \
    --vllm_gpu_memory_utilization 0.3
```
Environment variables:  
`export CUDA_VISIBLE_DEVICES=0`  
`export NCCL_P2P_LEVEL="NVL"`  
This worked as expected.

**Second Attempt (A800 x2):**  
Same command, but changed environment variable to:  
`export CUDA_VISIBLE_DEVICES=0,2`  

![Image](https://github.com/user-attachments/assets/f4829b23-1840-4be3-b1c7-1622eff78964)
The orange one is first train only train 45steps. But the reward is increased.

The multi-GPU training performed significantly worse. Repeated attempts with multi-GPU configurations all showed degraded results. The code remained unchanged.  

**Questions:**  
1. Is there a configuration error on my end?  
2. Could there be bugs in the trl library related to multi-GPU training?  
3. Other potential causes?  

I'm a beginner and would appreciate any guidance.

### System Info

A800 x4 80G GPUs

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by nopepper** [comment]
I've noticed a similar issue. DeepSpeed runs will eventually fail catastrophically and the reward plummets to 0. DDP doesn't work anymore when using `vLLM` and just hangs on the first step.

**Comment by qgallouedec** [comment]
You should set num_processes to 2 if you have 2 GPUs. 
Also, to get comparable results you must ensure that the effective batch size (num gpus*per device batch size*grad accum) is the same.
I'm not sure it will solve your problem, let me know

**Comment by Kfkcome** [comment]
> You should set num_processes to 2 if you have 2 GPUs. Also, to get comparable results you must ensure that the effective batch size (num gpus_per device batch size_grad accum) is the same. I'm not sure it will solve your problem, let me know

If I set num_processes = 1, the first gpu focus train and the second gpu focus sampling(vllm). 
And I use the same args, only change the `export CUDA_VISIBLE_DEVICES=0` to `export CUDA_VISIBLE_DEVICES=0,2`
I got the another test use the same args and one gpu.
The format reward increase very fast  just like the one before I test....  
 
![Image](https://github.com/user-attachments/assets/5ff946cc-b1a3-4608-9efa-6688c5078039)

**Comment by I-l-l-I** [comment]
@Kfkcome Something similar happened to me. There may be a communication problem between the GPUs. I solved it by replacing
```python
llm_model.load_weights(state_dict.items())
```
in `grpo_trainer._move_model_to_vllm` with
```python
llm_device = next(llm_model.parameters()).device
for name, param in state_dict.items():
    weight = param.to('cpu').to(llm_device)
    llm_model.load_weights(weights=[(name, weight)])
    del weight
```

**Comment by qgallouedec** [comment]
Hey @Kfkcome, sorry for the slow reply, we're catching up on the backlog.

The vLLM integration in GRPO has been fully redesigned since the version you were running. The weight-loading path that @I-l-l-I patched (`_move_model_to_vllm`) has also been substantially reworked. Closing as outdated, but please open a fresh issue if you reproduce a multi-GPU regression on a current TRL version
